### PR TITLE
Report individual findings rather than top-level file

### LINKED ISSUE #72: Report on individual files rather than only resultant aggregation at top level
Currently we are combining all nested findings as a single aggregated result for a top-level file on disk.

To provide transparency into exactly why a given top-level file is being flagged as vulnerable, we should report on individual files, even when nested, along with the full nesting path at which they can be found.

### PR DESCRIPTION
**Pull request by glynternet**
Closes: https://github.com/palantir/log4j-sniffer/issues/72

This PR changes the way vulnerabilities are reported when not in
`--file-path-only` mode.

Each finding is now reported individually, rather than reporting
and aggregation of all findings with only the top-level file on
disk.

For example, a vulnerable jar nested inside an archive will now be
reported with the vulnerability findings, rather than reporting on
the archive with an aggregation of all findings from within it.
Multiple vulnerable jars found within an archive will be reported
separately.

The path reported with a vulnerability finding is the full path
to a finding with archive layers delimited by a "!".
i.e. /path/to/archive!path/to/finding.jar shows that an archive
 at /path/to/archive contained a vulnerable jar at
 path/to/finding.jar within it.

---

To do this we now accept a `HandleFindingFunc` into the `Log4jIdentifier`, which is called for every finding encountered whilst crawling.

`Reporter.Collect` has also been renamed to `Report` to reflect better the behaviour that it has.

### CODE REVIEW

File: pkg/crawl/identify.go [hunk 1]
```diff
@@ -106,99 +107,159 @@ type Log4jIdentifier struct {
...
+	if len(nestedPaths) == 1 {
+		// we are on disk and so need to split using filepath to support different OS separators
+		_, filename := filepath.Split(path)
+		jarVersion, jarNameMatch = FileNameMatchesLog4jJar(filename)
+		jarFinding = JarName
+	} else {
+		// we are in an archive and so can assume separator is '/'
+		jarVersion, jarNameMatch = pathMatchesLog4JVersion(path)
+		jarFinding = JarNameInsideArchive
+		if jarNameMatch {
```
Comment (glynternet) [inline pkg/crawl/identify.go:200]: This behaviour is carried over from previous implementation.

@hpryce do we want to log at info level when we match _any_ `log4j-core` jar name or should we only do it when it is a vulnerable version?

Comment (hpryce) [reply pkg/crawl/identify.go:200]: Perhaps controversial, but I think we should drop vulnerable versions from identify.go entirely. We should just find everything here, then filter out in the reporting if it's not vulnerable. This probably means we need a new found but not vulnerable log line in the report.go for the human friendly output.

**Comment by nmiyake** [comment]
@glynternet haven't looked at this implementation closely, but if it basically just sets the output path to something of the form `/path/to/archive!path/to/finding.jar` and outputs that in the same way as top-level findings, then presumably this would be printed in the mode that prints file paths. We already have consumers that pipe that output to another process to them remove the impacted files, so this would be a change in behavior that could cause issues there.

Is there another field (like "details") where this would make more sense? I think it would be more consistent/understandable if the "path" always mapped to a file on disk. If we want further information like a path within an archive, perhaps that would make sense as an additional metadata field instead?

**Comment by hpryce** [comment]
This is a very good point @nmiyake, and one I hadn't included when I sketched this out with @glynternet. I think, in the name of avoiding needless breaks, we should add this as another field on the json object. The human readable output I think makes sense to use the paths as done here, and the file path listing should just list the top level path.

**Comment by glynternet** [comment]
Yup, very valid point and sloppy on my part for missing this.
I will update to make sure we are not making breaks to those existing workflows.

File: README.md [hunk 2]
```diff
@@ -128,13 +128,14 @@ a line of JSON.
 Here is an example of the output with `--json`:
 
 ```
-{"message":"CVE-2021-44832, CVE-2021-44228, CVE-2021-45046, CVE-2021-45105 detected","filePath":"examples/single_bad_version/log4j-core-2.14.1.jar","cvesDetected":["CVE-2021-44832","CVE-2021-44228","CVE-2021-45046","CVE-2021-45105"],"findings":["jndiLookupClassPackageAndName","jarName","jndiManagerClassPackageAndName","classFileMd5"],"log4jVersions":["2.14.0-2.14.1","2.14.1"]}
+{"message":"CVE-2021-44228, CVE-2021-44832, CVE-2021-45046, CVE-2021-45105 detected","filePath":"../examples/inside_a_dist/wrapped_log4j.tar.gz","detailedPath":"../examples/inside_a_dist/wrapped_log4j.tar.gz!log4j-core-2.14.1.jar","cvesDetected":["CVE-2021-44228","CVE-2021-44832","CVE-2021-45046","CVE-2021-45105"],"findings":["jndiLookupClassPackageAndName","jarNameInsideArchive","jndiManagerClassPackageAndName","classFileMd5"],"log4jVersions":["2.14.0-2.14.1","2.14.1"]}
```
Comment (glynternet) [inline README.md]:
```suggestion
{"message":"CVE-2021-44228, CVE-2021-44832, CVE-2021-45046, CVE-2021-45105 detected","filePath":"examples/inside_a_dist/wrapped_log4j.tar.gz","detailedPath":"examples/inside_a_dist/wrapped_log4j.tar.gz!log4j-core-2.14.1.jar","cvesDetected":["CVE-2021-44228","CVE-2021-44832","CVE-2021-45046","CVE-2021-45105"],"findings":["jndiLookupClassPackageAndName","jarNameInsideArchive","jndiManagerClassPackageAndName","classFileMd5"],"log4jVersions":["2.14.0-2.14.1","2.14.1"]}
```

File: README.md [hunk 3]
```diff
@@ -155,7 +156,7 @@ The summary outlines:
 - permissionDeniedErrors: the number of directories or files that could not be read due to permissions
 - pathErrors: the number of paths where an unexpected error occurred while trying to identify bad log4j versions
 - pathsSkipped: the numbers of paths skipped from full identification of bad log4j versions due to the config options set
-- numImpactedFiles: the total number of results previously output
+- findings: the total number of findings previously output
```
Comment (glynternet) [inline README.md]: Is this change of field going to be an issue?
I can separately report `numImpactedFiles` and `findings` if preferred.

Comment (hpryce) [reply README.md]: Yes, I think we should report separately. The number of impacted files is a better measure of work needing to be done, while findings just reflects how much there might be even if duplicated within the same archive.

Comment (glynternet) [reply README.md]: Cool, I've made that change.

**Comment by hpryce** [comment]
This looks good - the comment thread discussion around logging/handling of non-vulnerable versions is going to be handled in a follow up PR.

This now also keeps the json format the same.

**Comment by hpryce** [comment]
👍

**Pull request merged**